### ISSUE TITLE
Allow custom fields in profile object

### ISSUE THREAD

**Issue by ndom91** [nextauthjs/adapters#151]
Hey guys, wanted to open a new issue dedicated to this discussion as it was previously taking place in an already merged PR.. (https://github.com/nextauthjs/adapters/pull/132#issuecomment-869174077).

The currently implementation simply spreads in the entire `profile` object, possibly overwriting our preexisting fields such as `id`.

@sachinraja suggested a solution using a separate object for the custom user fields, such as:

```js
profile(profile) {
  return {
    name: profile.name ?? profile.login,
    email: profile.email,
    image: profile.avatar_url,
    __nextAuthCustomFields: {
      id: profile.id.toString(),
      username: profile.login,
      bio: profile.bio,
    },
  }
}
```

**Comment by anglinb** [comment]
Following up from the PR:
> So the provider id is overwriting our default primary key id field due to this spreading in all of their fields, right? 

Exactly!

>`@balazsorban44`, is it important for us to have the provider id? i.e. the github user id? What if we save that in a field called providerId or something along those lines?

The provider id already gets stored in the `accounts` table but not the `user` table, so you shouldn't have to worry about copying it here.

**Comment by sachinraja** [comment]
@anglinb I'm currently extending the GitHub provider like so:
```js
Providers.GitHub({
  clientId: envConfig.get('gitHub.clientId'),
  clientSecret: envConfig.get('gitHub.clientSecret'),
  profile(profile) {
    return {
      id: profile.id.toString(),
      username: profile.login,
      bio: profile.bio,
      name: profile.name ?? profile.login,
      email: profile.email,
      image: profile.avatar_url,
    }
  },
}),
```

The problem is that before the custom fields here, such as `username` were not being passed to `prisma.create.user`, but now they are since all properties are spread.

**Comment by anglinb** [comment]
@sachinraja Oohhhh ok that makes sense 👍 I also modified the adapter in a similar way to the provider to do additional setup in the create user function. If you're modifying the model you'll likely need to modify the adapter eventually? So maybe a good solution would be to make it easier to modify the adapter?

**Comment by sachinraja** [comment]
Modifying the adapter would work and is probably a better idea as using a special property feels hacky. Not sure if it's ideal for everyone though as it requires more from the user.

**Comment by anglinb** [comment]
Ya I'm on the same page. Maybe one way to make it easier would be to refactor how the initialization works so you could easily extend the adapter. 

Here's what I did but it would be way nicer if maybe `PrismaAdapter` too an additional parameter where you could implement the same functions that the base adapter does. The end result would be similar to how the providers works--you can easily override a specific part of the functionality, like you did with the `profile` function. 

```typescript

const CustomPrismaAdapter =  (prisma: PrismaClient) => {
  let originalAdpaterBuilder = PrismaAdapter(prisma);

  return {
    async getAdapter({ session, secret, ...appOptions }: { session: any, secret: string }) {
      // @ts-ignore
      let originalAdapter = await originalAdpaterBuilder.getAdapter({ session, secret, ...appOptions })
      const createUser = async (profile: any) => {
        // Create a default team 
        let personalTeam = await prisma.team.create({
          data:{
            name: profile.name,
            teamType: TeamType.PERSONAL,
          }
        })
        
        let { id, ...restOfProfile } = profile;
        return prisma.user.create({
          data: {
            name: profile.name,
            email: profile.email,
            image: profile.image,
            emailVerified: profile.emailVerified?.toISOString() ?? null,
            ...restOfProfile,
            personalTeamId: personalTeam.id,
          },
        })
      };

      let returning = { 
        ...originalAdapter,
        createUser
      };
      return returning
    }
  }
}

export default CustomPrismaAdapter;

```

**Comment by sachinraja** [comment]
Agreed, overrides would be useful. There is the [events callback](https://next-auth.js.org/configuration/events#createuser), but I don't think that overrides. What you've shown isn't too bad though, although it does look like the types need some work.

**Comment by anglinb** [comment]
Haha ya was lazy with the typing b/c I just wanted to get it working quickly 🤷  I'm happy enough with that approach to leave it for now but would happily upgrade to something cleaner if that's the direction ya'll take 👌

**Comment by sachinraja** [comment]
## Comparison
I've put together a simple table to hopefully help move this discussion along.

| Solutions  | Benefits |
| ------------- | ----------- |
| field in profile object    | <ul><li>less configuration for the user</li><li>the `createUser` function can be updated</li><li>no added TypeScript work</li></ul>  |
| extending the adapter | <ul><li>less "magic" as it does not rely on a custom field in the profile object</li><li>allows for consistency across adapters, no other adapter has a `customFields` property in the profile object</li></ul> |

**Comment by sachinraja** [comment]
@ndom91 @balazsorban44 Is there any way we can move this forward?

**Comment by ndom91** [comment]
Yeah so my vote would go to extending the adapter as we do want to expand custom fields to the rest of the adapters if one of them has it. To keep them all as close to one another feature wise as possible. 

What do you think @balazsorban44

**Comment by balazsorban44** [comment]
So I don't think we should add a custom field either, it feels weird. I am more into the idea of making it possible to override certain adapter methods, or somehow reuse the profile callback for the adapter, so it is "inferred" if that makes sense. Unfortunately I don't think we can squeeze that into 3.x as there is a lot of stuff going on already. I also want to change the adapter method signatures/merge some methods for less db querying, so it would make sense to wait until v4.

**Comment by sachinraja** [comment]
Overriding adapter methods sounds much nicer. If this can only be implemented in v4, I think #132 should be reverted.

> I think maybe the whole adapter should be a single class actually, that extends the base adapter.

Definitely would be easier to override by extending a class rather than [resorting to object spreads](https://github.com/nextauthjs/adapters/issues/151#issuecomment-869240764).

**Comment by balazsorban44** [comment]
We reverted the destructuring change for now, and will try to add it back and create a better adapter API for v4 with more clear messaging.
